### PR TITLE
Stenoish graders

### PR DESCRIPTION
**Pull request by NatTuck**
- Move to backburner for background jobs. This requires beanstalkd, but makes background jobs "instant" by being a blocking queue rather than polling.
- Sandbox graders use a new style of grading script, giving instructors more control over the build and test process. No more Makefiles unless you call "make" yourself.

This is an improvement by itself, but it's also prep for a move sandbox grading to external servers with "steno".

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -38,8 +38,8 @@ gem 'momentjs-rails', '>= 2.9.0' # needed for human-friendly textual dates
 gem 'font-awesome-rails'
 
 # needed to run graders in the background
-gem 'delayed_job', github: 'dsander/delayed_job', branch: 'rails51' # FIXME: github
```
Comment (blerner) [inline Gemfile]: I'm worried about deleting delayed_job entirely, this soon before the semester.  Is it possible to incorporate both for now, so we can migrate gently from one to the other?

File: app/models/grader.rb [hunk 2]
```diff
@@ -21,18 +23,29 @@ def self.unique
...
+
+    assignment = Assignment.find(asn_id)
+    submission = Submission.find(sub_id)
+
     ans = do_grading(assignment, submission)
     submission.compute_grade! if submission.grade_complete?
-    ans
+  end
+
+  def grade(assignment, submission, prio = 0)
```
Comment (blerner) [inline app/models/grader.rb]: IIRC, `grade` can be called in settings where I'm trying to find out "what is the grade of this submission for this assignment?", so returning no value (as removed on line 27 above) is unhelpful.  Why the change?

File: app/models/grader.rb [hunk 3]
```diff
@@ -21,18 +23,29 @@ def self.unique
     select(column_names - ["id"]).distinct
   end
 
-  def grade(assignment, submission)
+  def really_grade(asn_id, sub_id)
+    puts "XX really grade"
```
Comment (blerner) [inline app/models/grader.rb]: Stray `puts` comments?

File: app/models/graders/sandbox_grader.rb [hunk 4]
```diff
@@ -64,42 +67,26 @@ def run_sandbox(assignment, sub, sandbox, cont)
...
     parts = stdout.split("#{secret}\n")
     g.score = 0
     g.out_of = self.avail_score
 
     details_log = grader_dir.join("details.log")
     makefile_tap = grader_dir.join("makefile.tap")
+
     File.open(details_log, "w") do |details|
       details.write "== stdout ==\n\n#{stdout}\n\n== stderr ==\n\n#{stderr}\n\n== end of output ==\n"
+      puts "XX\n== stdout ==\n\n#{stdout}\n\n== stderr ==\n\n#{stderr}\n\n== end of output ==\n"
```
Comment (blerner) [inline app/models/graders/sandbox_grader.rb]: Stray `puts`?

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -63,5 +63,17 @@ window.current_page_name = "<%= params[:controller] %>/<%= params[:action] %>";
 
 <%= render "layouts/footer" %>
 
+<% if @queue_stats %>
```
Comment (blerner) [inline app/views/layouts/application.html.erb]: What is this, and how is it different from the /status page?  (That page, I think, holds more information about whose jobs are being graded...)

File: db/schema.rb [hunk 6]
```diff
@@ -41,6 +41,16 @@
     t.index ["course_id"], name: "index_assignments_on_course_id"
   end
 
+  create_table "codereview_matchings", force: :cascade do |t|
```
Comment (blerner) [inline db/schema.rb]: This is definitely going to cause merge conflicts.  Can you rebase on the latest changes, and re-number whatever migrations were needed, so that there's a linearizable history for the schema?

File: db/migrate/20170805001646_drop_delayed_jobs.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+class DropDelayedJobs < ActiveRecord::Migration[5.1]
+  def up
+    drop_table "delayed_jobs"
+  end
+
+  def down
+    raise ActiveRecord::IrreversibleMigration
```
Comment (blerner) [inline db/migrate/20170805001646_drop_delayed_jobs.rb]: Does this really need to be irreversible?  The contents of the delayed_jobs table are ephemeral anyway...  And having irreversible migrations makes it *really* hard to test and upgrade the database when needed.

File: lib/container.rb [hunk 8]
```diff
@@ -51,12 +52,48 @@ def chmod_r(path, mode)
...
-    command = %Q{lxc exec "#{@name}" -- bash -c \
-                   '(cd && BN_KEY="#{secret}" BN_SUB="#{sub}" BN_GRA="#{gra}" \
-                   ruby -I /tmp/bn/lib bn_driver.rb)'}
-    puts "Run driver: #{command}"
+  def exec_driver(secret, sub_up, gra_up, xtr_up = nil)
+    puts "XX In exec_driver"
+
+    FileUtils.mkdir_p("/tmp/bn")
+    Tempfile.create("driver.", "/tmp/bn") do |drv|
+      puts "XX in block"
```
Comment (blerner) [inline lib/container.rb]: What's with these "in block" comments, and where did block 3 go, and why can't they all go away?

File: sandbox/create-base.sh [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+apt-get upgrade -y
+
+apt-get install -y ruby python3
+apt-get install -y libarchive-zip-perl libipc-system-perl
+apt-get install -y clang valgrind build-essential
+apt-get install -y openjdk-8-jdk
+apt-get install -y libfuse-dev pkg-config
+apt-get install -y wamerican libbsd-dev libgmp-dev
+
+#apt-get install -y ocaml ocaml-native-compilers camlp4-extra opam \
```
Comment (blerner) [inline sandbox/create-base.sh]: Why are these commented out?  Might as well find out if these grader infrastructure can handle these programs too.

File: sandbox/driver.pl.erb [hunk 10]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/perl
```
Comment (blerner) [inline sandbox/driver.pl.erb]: It terrifies me that you can put embedded ruby inside perl...

File: sandbox/examples/demo/_grading/steno.rb [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+  def shell(cmd)
+    cmd.gsub!(%q{"}, %q{\"})
+    puts cmd
+    system(%Q{timeout -k #{@ktime} #{@timeout} bash -c "#{cmd}"}) or
+      raise Exception.new("shell(#{cmd}) -> #{$?}")
+  end
+
+  def unpack
+    if @sub =~ /\.zip$/i
```
Comment (blerner) [inline sandbox/examples/demo/_grading/steno.rb]: This is horribly unsafe, as I found out to my dismay.  There are no resource limits on this, no directory restrictions, etc.  I've written an ArchiveUtils module that unzips/untars/ungzips safely -- you should probably use that instead.

File: sandbox/examples/demo/_grading/steno.rb [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+      if hash0 != hash1
+        puts "Hash mismatch for _grading/#{name}"
+        puts "Hax!"
+        exit(1)
+      end
+    end
+  end
+
+  def shell(cmd)
+    cmd.gsub!(%q{"}, %q{\"})
```
Comment (blerner) [inline sandbox/examples/demo/_grading/steno.rb]: This is only correct for one level of nested quoting, right?

File: test/factories.rb [hunk 13]
```diff
@@ -78,9 +78,9 @@
     after(:build) do |assn|
       assn.course = assn.teamset.course
     end
-    after(:create) do |assn, evaluator|
-      assn.graders << create(:grader, assignment: assn)
-    end
+    #after(:create) do |assn, evaluator|
```
Comment (blerner) [inline test/factories.rb]: I think this breaks an invariant that assignments must have > 0 graders; this shouldn't be commented out.

File: test/models/sandbox_test.rb [hunk 14]
```diff
@@ -9,11 +9,12 @@ class SandboxTest < ActiveSupport::TestCase
   test "run a sandbox grader lazy mode" do
     skip unless ENV['TEST_SANDBOX']
 
-    assign = create(:assignment, type: "files", course: @cs101, name: "Test Assignment")
+    assign = create(:assignment, type: "Files", course: @cs101, name: "Test Assignment")
     grdtar = Rails.root.join('test', 'fixtures', 'files', 'TestScript', 'test.pl')
     upload = simulated_upload(@fred, grdtar)
     upload.save!
-    grdcfg = build(:grader, type: "SandboxGrader", upload: upload, params: "lazy.rb", assignment: assign, order: 1)
+    grdcfg = build(:grader, type: "SandboxGrader", upload: upload, params: "lazy.rb",
```
Comment (blerner) [inline test/models/sandbox_test.rb]: Isn't this defunct now?  What's a lazy sandbox grader?

File: app/models/grader.rb (hunk 3 above)
Comment (NatTuck) [inline app/models/grader.rb]: Yup.

File: Gemfile (hunk 1 above)
Comment (NatTuck) [inline Gemfile]: AFAIK we use background jobs for two things: 1. graders 2. sending email.

Sending email goes through ActiveJob, where backburner is a drop in replacement for delayed_job. Conceptually we could stick with delayed_job here, but that seems unnecessarily redundant.

For graders, we shouldn't have two different methods of delaying jobs.

Comment (blerner) [inline Gemfile]: I'm pretty sure BN can't send email at the moment, actually.  I agree we shouldn't have two methods of delaying jobs, but I'm leery of switching this over right before the semester.

Question: is switching to backburner orthogonal to using Steno as the sandboxing mechanism?  If so, then that should be a separate PR so I can disentangle them in my head...

File: app/models/grader.rb (hunk 2 above)
Comment (NatTuck) [inline app/models/grader.rb]: I was wondering about the synchronous calls. Those don't work - running a grading script may take a long time, and we don't want to be blocking a web request on it. Those call sites should be modified to do a field access and accept "not done yet" as an answer.

Comment (blerner) [inline app/models/grader.rb]: The primary use case for this is when I as a professor need to regrade a student's submission immediately, and want to bypass the job queue.

Comment (blerner) [inline app/models/grader.rb]: And...the synchronous nature of it is ok to me, since I'm often trying to identify "is this thing hanging when I don't expect it to be hanging?"  So the web-request timeout, and the blocking of everyone else while I'm trying to diagnose a problem, is fine by me.

**Comment by NatTuck** [comment]
This needs one more commit to be feature-complete, and a pass to remove extra printing.

File: lib/container.rb (hunk 8 above)
Comment (NatTuck) [inline lib/container.rb]: "XX" is my search term for prints that I should have deleted.

File: sandbox/create-base.sh (hunk 9 above)
Comment (NatTuck) [inline sandbox/create-base.sh]: Because they're really slow to install.

File: sandbox/driver.pl.erb (hunk 10 above)
Comment (NatTuck) [inline sandbox/driver.pl.erb]: Don't worry. The perl executes embedded shell code.

File: sandbox/examples/demo/_grading/steno.rb (hunk 11 above)
Comment (NatTuck) [inline sandbox/examples/demo/_grading/steno.rb]: We can pull in ArchiveUtils to the sample _grading directory.

Do you have a malicious zip archive to test with? I'd like to make it so that the sandbox catches that error - although we'll still want ArchiveUtils so we can report clean errors.

File: test/models/sandbox_test.rb (hunk 14 above)
Comment (NatTuck) [inline test/models/sandbox_test.rb]: Yea, these test are all defunct now.

Tests for sandbox graders are annoying, since they can't run on Travis and therefore break 100% of the time. That's one of the advantages to the steno plan - if sandbox grading is external, we can keep the Bottlenose repo 100% travis testable.

File: sandbox/examples/demo/_grading/steno.rb (hunk 12 above)
Comment (NatTuck) [inline sandbox/examples/demo/_grading/steno.rb]: Probably not even then. Bash really needs ruby/perl style nestable quotes.

File: app/views/layouts/application.html.erb (hunk 5 above)
Comment (NatTuck) [inline app/views/layouts/application.html.erb]: Probably just a stray print, but maybe a useful UI element for students.

Interestingly, it's useful in this particular version but not in the external-service version.

File: db/schema.rb (hunk 6 above)
Comment (NatTuck) [inline db/schema.rb]: We should probably land your PR first anyway. I need npm for the steno websocket integration.

File: Gemfile (hunk 1 above)
Comment (NatTuck) [inline Gemfile]: Steno (the other repo) provides real-time in browser output for grading scripts. This needs backburner on the rails side for instant async execution.

This pull request could be conceptually split into (use the new grading script protocol) and (move to backburner), which are basically orthogonal, but both pieces are needed for Steno.

File: app/models/grader.rb (hunk 2 above)
Comment (NatTuck) [inline app/models/grader.rb]: That's ugly. It kind of works with the non-sandboxed style checkers, but it'll be really bad with sandboxes. 

And with Steno, I don't even have a plan for synchronous calls. Currently, results get POSTed back later.

File: db/migrate/20170805001646_drop_delayed_jobs.rb (hunk 7 above)
Comment (NatTuck) [inline db/migrate/20170805001646_drop_delayed_jobs.rb]: Nah. I can have the down method recreate the table.

**Comment by NatTuck** [comment]
New grading protocol:

A grading job has three associated files:
 - A submission
 - A grading archive
 - (optional) An extra file

Grading proceeds as follows:
 - The files are downloaded into /home/ubuntu in the sandbox.
 - The grading archive is unpacked.
 - We run the _grading/grade.rb script from the grading archive.

When _grading/grade.rb runs, it gets five environment variables:
 - GRA = name of the grading archive
 - SUB = name of the submission file
 - XTR = name of the extra file, if any
 - COOKIE = a secret string, to delimit TAP output
 - TIMEOUT = max time in seconds for a single subprocess

The grading script does whatever it wants with that information, as long as it outputs some TAP text to stdout delimited by the COOKIE.

The extra file is for parameterization to implement specific graders. For JUnitGrader, you'd build a junit-grading.tar.gz archive that expects a the JUnit tests file as the XTR parameter and then hard code that as the grading archive in the Bottlenose-side Grader implementation.

File: test/factories.rb (hunk 13 above)
Comment (NatTuck) [inline test/factories.rb]: Hmm... I can't do the thing that was breaking anyway.

File: sandbox/examples/demo/_grading/steno.rb (hunk 11 above)
Comment (blerner) [inline sandbox/examples/demo/_grading/steno.rb]: Yes, I do -- several varieties (bad symlinks, ginormous files, bad relative paths, and different archive formats).  But they're (duh!) large files, so it'll take a bit to send them to you.  I'll get it to you next week when I'm back on campus.

File: Gemfile (hunk 1 above)
Comment (blerner) [inline Gemfile]: Ok.  Is it correct to say that the *existing* {JUnit, JavaStyle, Checker, RacketStyle}Graders will all continue to work as they currently are, with this change?  Or do all of those need to be rewritten concurrently with this change, to use Steno as the backend?  I'm *really* hesitant to do that right now -- I'd rather you beta-test Steno with your courses this fall (where you can babysit it and keep track of any necessary changes) than with Fundies (where I probably won't have any bandwidth to do the same).

File: db/schema.rb (hunk 6 above)
Comment (blerner) [inline db/schema.rb]: Just made my PR; couldn't do it last week while I was away...

File: sandbox/create-base.sh (hunk 9 above)
Comment (blerner) [inline sandbox/create-base.sh]: But doesn't this script only get run once per BN installation?

File: sandbox/driver.pl.erb (hunk 10 above)
Comment (blerner) [inline sandbox/driver.pl.erb]: yodawg...

**Comment by blerner** [comment]
Ok, some minor remaining comments (e.g. about readability of the Status page), and a bit more confusion on my part about how to rewrite the existing grader types to take advantage of this new framework.  When are you on campus to review these two PRs and merge them?

File: sandbox/examples/extra/_grading/grade.rb [hunk 15]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env ruby
+
+require 'steno'
+
+fs = ['Makefile', 'test.pl']
+
+FileUtils.cp("test.pl", "_grading")
```
Comment (blerner) [inline sandbox/examples/extra/_grading/grade.rb]: Should be something like 

    FileUtils.cp(ENV["XTR"], "_grading/test.pl")

To demonstrate proper parameterization. Check driver.pl.erb for details.
